test/framework/test_options.go

```diff
@@ -0,0 +1,118 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"context"
+	"github.com/onsi/ginkgo"
+	"time"
+)
+
+// Test options contains options to add additional functionality
+// or cleanup handlers to a testcase.
+type TestOptions struct {
+	// AfterTests holds a list of all registered AfterTest functions
+	// that are executed when the test has finished.
+	AfterTests []func()
+
+	// AfterTests holds a list of all registered contextified AfterTest functions
+	// that are executed when the test has finished.
+	CAfterTests []CAfterTestOptions
+}
+
+// ApplyOptions applies the given test options on these options.
+func (o *TestOptions) ApplyOptions(opts []TestOption) *TestOptions {
+	for _, opt := range opts {
+		opt.ApplyToTest(o)
+	}
+	return o
+}
+
+// Complete registers all test options that are configured.
+// it should be a function that configures a ginkgo test case
+// This should get called when all options are applied.
+func (o *TestOptions) Complete(it func()) {
+
+	if len(o.AfterTests) == 0 && len(o.CAfterTests) == 0 {
+		it()
+		return
+	}
+
+	// Create a new context so that the AfterTests function only runs once after this one testcase.
+	// Otherwise the after tests would run after every test case in the outer context.
+	ginkgo.Context("", func() {
+		it()
+
+		for _, aftertest := range o.AfterTests {
+			// register afterTest to global aftersuite in case the test interrupts
+			h := AddCleanupAction(aftertest)
+			ginkgo.AfterEach(func() {
+				RemoveCleanupAction(h)
+				aftertest()
+			})
+		}
+
+		for _, caftertest := range o.CAfterTests {
+			// register afterTest to global aftersuite in case the test interrupts
+			h := AddCleanupAction(func() {
+				contextify(caftertest.Body, caftertest.Timeout)()
+			})
+			CAfterEach(func(ctx context.Context) {
+				RemoveCleanupAction(h)
+				caftertest.Body(ctx)
+			}, caftertest.Timeout)
+		}
+	})
+}
+
+// CAfterTestOptions contains options for contextified after test function.
+type CAfterTestOptions struct {
+	Body    func(ctx context.Context)
+	Timeout time.Duration
+}
+
+// ApplyToTest adds contextified after test functions to test options
+func (at *CAfterTestOptions) ApplyToTest(opts *TestOptions) {
+	opts.CAfterTests = append(opts.CAfterTests, *at)
+}
+
+// TestOption is some configuration that modifies options for testcase.
+type TestOption interface {
+	// ApplyToTest applies this configuration to the given test options.
+	ApplyToTest(*TestOptions)
+}
+
+// AfterTests are functions that should run when a test has finished
+type AfterTests []func()
```

Comment:
Can we keep these private?
```suggestion
type afterTests []func()
```